linkerd/app/outbound/src/ingress.rs

```diff
@@ -0,0 +1,204 @@
+                    SpanConverter::server(span_sink, trace_labels())
+                })))
+                .push(metrics.stack.layer(stack_labels("source")))
+                .push_failfast(dispatch_timeout)
+                .push_spawn_buffer(buffer_capacity)
+                .box_http_response(),
+        )
+        .check_new_service::<http::Accept, http::Request<_>>()
+        .push(svc::layer::mk(http::normalize_uri::MakeNormalizeUri::new))
+        .check_new_service::<http::Accept, http::Request<_>>()
+        .instrument(|a: &http::Accept| info_span!("http", v = %a.protocol))
+        .push_map_target(http::Accept::from)
+        .check_new_service::<(http::Version, tcp::Accept), http::Request<_>>()
+        .into_inner();
+
+    let tcp = svc::stack(tcp)
+        .push_map_target(tcp::Endpoint::from_accept(
+            tls::ReasonForNoPeerName::IngressNonHttp,
+        ))
+        .into_inner();
+
+    svc::stack(http::DetectHttp::new(h2_settings, http, tcp, drain))
+        .check_new_service::<tcp::Accept, io::PrefixedIo<transport::metrics::SensorIo<I>>>()
+        .push_on_response(svc::layers().push_spawn_buffer(buffer_capacity).push(
+            transport::Prefix::layer(
+                http::Version::DETECT_BUFFER_CAPACITY,
+                detect_protocol_timeout,
+            ),
+        ))
+        .check_new_service::<tcp::Accept, transport::metrics::SensorIo<I>>()
+        .push(metrics.transport.layer_accept())
+        .push_map_target(tcp::Accept::from)
+        .check_new_service::<listen::Addrs, I>()
+        .into_inner()
+}
+
+#[derive(Clone)]
+struct AllowHttpProfile(AddrMatch);
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+struct Target {
+    dst: Addr,
+    accept: http::Accept,
+}
+
+#[derive(Clone)]
+struct TargetPerRequest(http::Accept);
+
+// === AllowHttpProfile ===
+
+impl svc::stack::FilterRequest<Target> for AllowHttpProfile {
+    type Request = Addr;
+
+    fn filter(&self, Target { dst, .. }: Target) -> Result<Addr, Error> {
+        if self.0.matches(&dst) {
+            Ok(dst)
+        } else {
+            Err(discovery_rejected().into())
+        }
+    }
+}
+
+// === impl Target ===
+
+impl From<(Option<profiles::Receiver>, Target)> for http::Logical {
+    fn from((p, Target { accept, .. }): (Option<profiles::Receiver>, Target)) -> Self {
+        Self {
+            profile: p,
+            orig_dst: accept.orig_dst,
+            protocol: accept.protocol,
+        }
+    }
+}
+
+// === TargetPerRequest ===
+
+impl TargetPerRequest {
+    fn accept(a: http::Accept) -> Self {
+        Self(a)
+    }
+}
+
+impl<B> svc::stack::RecognizeRoute<http::Request<B>> for TargetPerRequest {
+    type Key = Target;
+
+    fn recognize(&self, req: &http::Request<B>) -> Self::Key {
+        Target {
+            accept: self.0,
+            dst: http_request_l5d_override_dst_addr(req).unwrap_or_else(|_| self.0.orig_dst.into()),
```

Comment:
nit/TIOLI: it might be worth having a debug event indicating whether or not we saw the override header, or something, since this is kinda load-bearing?

Reply:
It should be pretty obvious based on the resulting target, which is ends up in span metadata